awscli/utils.py

```diff
@@ -17,10 +17,138 @@
 import os
 import sys
 import subprocess
+import logging
 
 from awscli.compat import six
 from awscli.compat import get_binary_stdout
 from awscli.compat import get_popen_kwargs_for_pager_cmd
+from botocore.utils import IMDSFetcher
+from botocore.configprovider import BaseProvider
+from botocore.configprovider import InstanceVarProvider
+from botocore.configprovider import EnvironmentProvider
+from botocore.configprovider import ScopedConfigProvider
+from botocore.configprovider import ConstantProvider
+from botocore.configprovider import ChainProvider
+
+logger = logging.getLogger(__name__)
+
+
+class IMDSRegionProvider(BaseProvider):
+    def __init__(self, session, environ=None, fetcher=None):
+        """Initialize IMDSRegionProvider.
+
+        :type session: :class:`botocore.session.Session`
+        :param session: The session is needed to look up configuration for
+            how to contact the instance metadata service. Specifically the
+            whether or not it should use the IMDS region at all, and if so how
+            to configure the timeout and number of attempts to reach the
+            service.
+
+        :type environ: None or dict
+        :param environ: A dictionary of environment variables to use. If
+            ``None`` is the argument then ``os.environ`` will be used by
+            default.
+
+        :type fecther: :class:`botocore.utils.InstanceMetadataRegionFetcher`
+        :param fetcher: The class to actually handle the fetching of the region
+            from the IMDS. If not provided a default one will be created.
+        """
+        self._session = session
+        if environ is None:
+            environ = os.environ
+        self._environ = environ
+        self._fetcher = fetcher
+
+    def provide(self):
+        """Provide the region value from IMDS."""
+        instance_region = self._get_instance_metadata_region()
+        return instance_region
+
+    def _get_instance_metadata_region(self):
+        # First we construct a configuration chain to fetch the config value
+        # cli_use_instance_region to determine whether or not to use the
+        # instance's region as a last fallback if we do not have the region
+        # configured. By default this is turned on for convenience.
+        should_use_instance_region = \
+            self._construct_use_instance_region_chain().provide()
+        if not should_use_instance_region.lower() == 'true':
+            return None
+
+        fetcher = self._get_fetcher()
+        region = fetcher.retrieve_region()
+        return region
+
+    def _construct_use_instance_region_chain(self):
```

Comment:
Talked offline about this, but we can hold off on this for now. There is not really too much benefit in specifically opting out of this because you can opt-out by setting a region since it is last in the chain and if you really did not want to hit IMDS you would just set the environment variable to turn it off completely.